### PR TITLE
#21: Adding tags and scores storages

### PR DESCRIPTION
**Pull request by kirillt**
These issues are solved in this PR:
- #21
- #47

New `BadStorageFile` exception must be handled by apps instead of `isCorrupted` predicate.

`PreviewStorage` and `MetadataStorage` are refactored to conform to new `Processor` interface. `Processor` objects are similar to `Storage` but also have mandatory `updates: Flow<Update>` field, which can be used for subscription. In future, it should be good idea to refactor them into pipeline-like structures with explicit input and output.

Internally, `PreviewProcessor` and `MetadataProcessor` objects contain storages of type `FolderStorage` inside.

Interface `Storage` doesn't have subscription mechanism at this moment.

- [x] After each reindex all files in storages are overwritten
- [x] ~~Fast change of scores causes `Timestamp didn't update` illegal state~~
- [x] Verify that JSONs are persisted as plain text.
- [x] Verify that persistence of the data is happening after UI becomes interactive.
- [x] Create generic `Aggregation` class suitable for storages instead of numerous similar `Aggregated*Storage` classes
- [x] Aggregations should not require user to provide path to the resource. `ResourceId` must be universal identifier.
- [x] ~~Sometimes `Dispatchers.IO` used when `Dispatchers.Default` should be used.~~
       `Dispatchers.IO` isn't single-threaded as I thought before.
- [x] ~~Function `readFromDisk` need to be run in a coroutine, but we cannot mark it as `suspend fun`.~~
       ~~Function `writeToDisk` also needs to be run in a coroutine.~~
- [x] Additional index cache in `ResourceIndexRepo`

### CODE REVIEW

File: lib/src/main/java/space/taran/arklib/domain/preview/BitmapStorage.kt [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+            Preview.COMPRESSION_QUALITY,
+            stream
+        )
+
+        return stream.toByteArray()
+    }
+
+    override fun valueFromBinary(raw: ByteArray): Bitmap {
+        return Glide.with(app).asBitmap().load(raw).submit().get()
+    }
```
[the remark this answers is not in the document]
Comment (kirillt) [reply lib/src/main/java/space/taran/arklib/domain/preview/BitmapStorage.kt:29]: Good point

[the remark this answers is not in the document]
Comment (kirillt) [reply lib/src/main/java/space/taran/arklib/domain/preview/BitmapStorage.kt:30]: Actually, we should never use these binary conversions for `Image` kind of resources. Only pdfs, videos and thumbnails must use `BitmapStorage`. Path to original image should always be retrieved from somewhere else.

[the remark this answers is not in the document]
Comment (kirillt) [reply lib/src/main/java/space/taran/arklib/domain/preview/BitmapStorage.kt:30]: The path to original image was taken from `index` instance passed to `PreviewStorage`.

I've refactored it to create tiny cache field with paths to resources of `IMAGE` kind.
The cache is written on each update from `MetadataStorage`.

File: lib/build.gradle [hunk 2]
```diff
@@ -96,7 +96,7 @@ publishing {
         debug(MavenPublication) {
             groupId = "space.taran"
             artifactId = "arklib-debug"
-            version = libVersion
+            version = '999'
```
Comment (kirillt) [inline lib/build.gradle:99]: [debug]

File: lib/src/main/java/space/taran/arklib/domain/meta/RootMetadataProcessor.kt [hunk 3]
```diff
@@ -28,12 +27,30 @@ class RootMetadataProcessor(
     init {
         Log.i(LOG_PREFIX, "Initializing metadata storage for root $root")
 
-        storage.refresh()
+        storage.init()
+        // todo: seems to be not needed: storage.refresh()
```
Comment (kirillt) [inline lib/src/main/java/space/taran/arklib/domain/meta/RootMetadataProcessor.kt:31]: [todo]

File: lib/src/main/java/space/taran/arklib/domain/preview/RootPreviewProcessor.kt [hunk 4]
```diff
@@ -13,7 +13,7 @@ import java.nio.file.Path
 
 class RootPreviewProcessor(
     private val scope: CoroutineScope,
-    private val index: RootIndex,
+    private val index: RootIndex, //todo can we get rid of index dependency in processors?
```
Comment (kirillt) [inline lib/src/main/java/space/taran/arklib/domain/preview/RootPreviewProcessor.kt:16]: [todo]

File: lib/src/main/java/space/taran/arklib/domain/preview/RootPreviewProcessor.kt [hunk 5]
```diff
@@ -32,8 +32,10 @@ class RootPreviewProcessor(
     init {
         Log.i(LOG_PREFIX, "Initializing previews storage for root $root")
 
-        previews.refresh()
-        thumbnails.refresh()
+        previews.init()
+        thumbnails.init()
+        //todo: seems to be not needed: previews.refresh()
+        //todo: seems to be not needed: thumbnails.refresh()
```
Comment (kirillt) [inline lib/src/main/java/space/taran/arklib/domain/preview/RootPreviewProcessor.kt:38]: [todo]

File: lib/src/main/java/space/taran/arklib/domain/preview/RootPreviewProcessor.kt [hunk 6]
```diff
@@ -42,10 +44,8 @@ class RootPreviewProcessor(
         initUpdatedResourcesListener()
     }
 
-    override fun forget(id: ResourceId) {
-        previews.remove(id)
-        thumbnails.remove(id)
-    }
+    // todo: could be easier for presenters initialization
```
Comment (kirillt) [inline lib/src/main/java/space/taran/arklib/domain/preview/RootPreviewProcessor.kt:47]: [todo]

File: lib/src/main/java/space/taran/arklib/domain/preview/RootPreviewProcessor.kt [hunk 7]
```diff
@@ -101,14 +106,31 @@ class RootPreviewProcessor(
         scope.launch(Dispatchers.IO) {
             _busy.emit(true)
 
+            metadata.state().forEach { (id, meta) ->
+                val path = index.getPath(id)!! //todo: better to get rid of index dependency
```
Comment (kirillt) [inline lib/src/main/java/space/taran/arklib/domain/preview/RootPreviewProcessor.kt:110]: [todo]

File: lib/src/main/java/space/taran/arklib/domain/preview/RootPreviewProcessor.kt [hunk 8]
```diff
@@ -101,14 +106,31 @@ class RootPreviewProcessor(
...
+                _busy.emit(true)
+
                 when (update) {
                     is MetadataUpdate.Added -> generate(update)
                     is MetadataUpdate.Deleted -> forget(update.id)
                 }
-            }
 
-            _busy.emit(false)
+                //todo: we should batch updates
```
Comment (kirillt) [inline lib/src/main/java/space/taran/arklib/domain/preview/RootPreviewProcessor.kt:127]: [todo]

File: lib/src/main/java/space/taran/arklib/domain/processor/Processor.kt [hunk 9]
```diff
@@ -13,7 +13,9 @@ abstract class Processor<Value, Update> {
     protected val _busy = MutableStateFlow(false)
     val busy: StateFlow<Boolean> = _busy.asStateFlow()
 
-    abstract fun retrieve(id: ResourceId): Result<Value>
+    abstract fun state(): Map<ResourceId, Value>
+
+    abstract fun retrieve(id: ResourceId): Result<Value> //todo: probably should not be exposed
```
Comment (kirillt) [inline lib/src/main/java/space/taran/arklib/domain/processor/Processor.kt:18]: [todo]

File: lib/src/main/java/space/taran/arklib/domain/tags/TagStorageRepo.kt [hunk 10]
```diff
@@ -20,10 +20,15 @@ class TagsStorageRepo(private val scope: CoroutineScope) {
         }
     }
 
+    //todo is it worth always providing up-to-date storage?
+    //todo i.e. need to call `refresh` on cached values
```
Comment (kirillt) [inline lib/src/main/java/space/taran/arklib/domain/tags/TagStorageRepo.kt:24]: [todo]

File: lib/src/main/java/space/taran/arklib/domain/tags/TagStorageRepo.kt [hunk 11]
```diff
@@ -20,10 +20,15 @@ class TagsStorageRepo(private val scope: CoroutineScope) {
...
+
     fun provide(root: RootIndex): RootTagsStorage =
         storageByRoot[root.path] ?: RootTagsStorage(
             scope, root.path,
         ).also {
+            it.init()
+            it.init() //todo test
```
Comment (kirillt) [inline lib/src/main/java/space/taran/arklib/domain/tags/TagStorageRepo.kt:31]: [debug]

File: lib/src/main/java/space/taran/arklib/domain/processor/AggregateProcessor.kt [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+): Processor<Value, Update>() {
+
+    init {
+        fun atLeastOneShardIsBusy() = shards.map { it.busy.value }.contains(true)
+
+        shards.forEach { shard ->
+            shard.busy.onEach {
+                _busy.emit(atLeastOneShardIsBusy())
+            }
+        }
```
Comment (mdrlzy) [inline lib/src/main/java/space/taran/arklib/domain/processor/AggregateProcessor.kt:20]: `.launchIn(appScope)`

File: lib/src/main/java/space/taran/arklib/domain/score/ScoreStorageRepo.kt [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+
+import kotlinx.coroutines.CoroutineScope
+import space.taran.arklib.domain.index.ResourceIndex
+import space.taran.arklib.domain.index.RootIndex
+import java.nio.file.Path
+
+class ScoreStorageRepo(private val scope: CoroutineScope) {
+    private val storageByRoot = mutableMapOf<Path, RootScoreStorage>()
+
+    suspend fun provide(index: ResourceIndex): Any {
```
Comment (mdrlzy) [inline lib/src/main/java/space/taran/arklib/domain/score/ScoreStorageRepo.kt:11]: `Any` is bad practice, isn't there a suitable interface?

File: lib/src/main/java/space/taran/arklib/domain/tags/TagStorageRepo.kt [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+
+import kotlinx.coroutines.CoroutineScope
+import space.taran.arklib.domain.index.ResourceIndex
+import space.taran.arklib.domain.index.RootIndex
+import java.nio.file.Path
+
+class TagsStorageRepo(private val scope: CoroutineScope) {
+    private val storageByRoot = mutableMapOf<Path, RootTagsStorage>()
+
+    suspend fun provide(index: ResourceIndex): Any {
```
Comment (mdrlzy) [inline lib/src/main/java/space/taran/arklib/domain/tags/TagStorageRepo.kt:11]: `Any` is bad practice

File: lib/src/main/java/space/taran/arklib/domain/storage/BaseStorage.kt [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+                setValue(addedId, it[addedId]!!)
+            }
+
+            for (knownId in it.keys.intersect(valueById.keys)) {
+                val theirs = it[knownId]!!
+                val ours = getValue(knownId)
+                if (theirs != ours) {
+                    //if `ours` or `theirs` was `null` they wouldn't conflict
+                    require(theirs != null)
+                    require(ours != null)
```
Comment (mdrlzy) [inline lib/src/main/java/space/taran/arklib/domain/storage/BaseStorage.kt:94]: Looks like `!!` operator is not needed.

File: lib/src/main/java/space/taran/arklib/domain/tags/TagStorageRepo.kt (hunk 14 above)
Comment (kirillt) [reply lib/src/main/java/space/taran/arklib/domain/tags/TagStorageRepo.kt:11]: Good catch, I didn't notice this.

File: lib/src/main/java/space/taran/arklib/domain/score/ScoreStorageRepo.kt (hunk 13 above)
Comment (kirillt) [reply lib/src/main/java/space/taran/arklib/domain/score/ScoreStorageRepo.kt:11]: It's unchecked copy-paste. Or something random. Must be fixed for sure.

**Comment by mdrlzy** [comment]
```
06-09 00:56:01.152 E/ACRA    (16329): ACRA caught a NullPointerException for space.taran.arknavigator
06-09 00:56:01.152 E/ACRA    (16329): java.lang.NullPointerException
06-09 00:56:01.152 E/ACRA    (16329): 	at space.taran.arklib.domain.preview.RootPreviewProcessor.initKnownResources(RootPreviewProcessor.kt:128)
06-09 00:56:01.152 E/ACRA    (16329): 	at space.taran.arklib.domain.preview.RootPreviewProcessor.access$initKnownResources(RootPreviewProcessor.kt:15)
06-09 00:56:01.152 E/ACRA    (16329): 	at space.taran.arklib.domain.preview.RootPreviewProcessor$init$2.invokeSuspend(RootPreviewProcessor.kt:44)
06-09 00:56:01.152 E/ACRA    (16329): 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
06-09 00:56:01.152 E/ACRA    (16329): 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
06-09 00:56:01.152 E/ACRA    (16329): 	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570)
06-09 00:56:01.152 E/ACRA    (16329): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
06-09 00:56:01.152 E/ACRA    (16329): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677)
06-09 00:56:01.152 E/ACRA    (16329): 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664)
06-09 00:56:01.152 E/ACRA    (16329): 	Suppressed: kotlinx.coroutines.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@a414215, Dispatchers.Default]
```

```
    private suspend fun initKnownResources() {
        _busy.emit(true)
        metadata.state().forEach { (id, meta) ->
            val path = index.getPath(id)!!
            generate(id, path, meta)
        }
    ...
    }
```

RootPreviewProcessor takes meta, but there is no such id in index, because resource was deleted, but meta was not deleted

**Comment by mdrlzy** [comment]
After each reindex all files in storages are overwritten

File: lib/src/main/java/space/taran/arklib/domain/preview/RootPreviewProcessor.kt [hunk 16]
```diff
@@ -61,69 +62,84 @@ class RootPreviewProcessor(
     override fun forget(id: ResourceId) {
         previews.remove(id)
         thumbnails.remove(id)
+        images.remove(id)
```
Comment (kirillt) [inline lib/src/main/java/space/taran/arklib/domain/preview/RootPreviewProcessor.kt:65]: We must not remove any of resources!

File: lib/src/main/java/space/taran/arklib/domain/meta/MetadataGenerator.kt [hunk 17]
```diff
@@ -31,17 +31,16 @@ interface MetadataGenerator {
             val generator = GENERATORS.find {
                 it.isValid(path)
             } ?: let {
-                return Result.failure(
-                    IllegalArgumentException("No generators found for $path")
-                )
+                Log.d(LOG_PREFIX, "No generators found for $path")
+                return Result.success(Metadata.Unknown())
```
Comment (kirillt) [inline lib/src/main/java/space/taran/arklib/domain/meta/MetadataGenerator.kt:35]: Am I right, if we add new kinds later, we could just invalidate all existing metadata to generate metadata for what is `UNKNOWN` now?

Comment (mdrlzy) [reply lib/src/main/java/space/taran/arklib/domain/meta/MetadataGenerator.kt:35]: Yep, we can do it

File: lib/src/main/java/space/taran/arklib/domain/preview/RootPreviewProcessor.kt [hunk 18]
```diff
@@ -125,8 +125,13 @@ class RootPreviewProcessor private constructor(
     private suspend fun initKnownResources() {
         _busy.emit(true)
         metadata.state().forEach { (id, meta) ->
-            val path = index.getPath(id)!!
-            generate(id, path, meta)
+            // Workaround
+            // Right now we are not removing lost resource meta, which causes npe
+            // Here should be index.getPath(id)!!, see:
+            // https://github.com/ARK-Builders/arklib-android/issues/70
```
Comment (kirillt) [inline lib/src/main/java/space/taran/arklib/domain/preview/RootPreviewProcessor.kt:131]: https://github.com/ARK-Builders/arklib-android/issues/70

File: lib/src/main/java/space/taran/arklib/domain/preview/RootPreviewProcessor.kt [hunk 19]
```diff
@@ -45,20 +45,20 @@ class RootPreviewProcessor private constructor(
...
-    // `locator.status` can be `ABSENT` and should be checked by consumer
-    override fun retrieve(id: ResourceId): Result<PreviewLocator> = metadata
-        .retrieve(id)
-        .map { metadata ->
-            val locator = if (metadata.kind == Kind.IMAGE) {
-                PreviewLocator(root, id, images[id])
-            } else {
-                PreviewLocator(root, id)
+    override fun retrieve(id: ResourceId): Result<PreviewLocator> {
+        val meta =  metadata.retrieve(id).getOrNull()
```
Comment (kirillt) [inline lib/src/main/java/space/taran/arklib/domain/preview/RootPreviewProcessor.kt:49]: Also `onFailure` could be used, but your method is also good.

File: lib/build.gradle [hunk 20]
```diff
@@ -81,7 +81,8 @@ tasks.whenTaskAdded { task ->
     }
 }
 
-def libVersion = scmVersion.version
+// def libVersion = scmVersion.version
+def libVersion = '999'
```
Comment (kirillt) [inline lib/build.gradle:85]: _Remove before merge_

**Review by mdrlzy — APPROVED** [review]

**Pull request merged**